hepstats/hypotests/calculators/frequentist_calculator.py

```diff
@@ -0,0 +1,425 @@
+
+    def toys_loss(self, parameter):
+        """
+        Construct a loss function constructed with a sampler for a given floating parameter
+
+        Args:
+            parameter: floating parameter in the sampler
+        Returns:
+             Loss function
+
+        Example with `zfit`:
+            >>> loss = calc.toys_loss(zfit.Parameter("mean"))
+        """
+        if parameter.name not in self._toys_loss:
+            sampler = self.sampler(floating_params=[parameter])
+            self._toys_loss[parameter.name] = self.lossbuilder(self.model, sampler)
+        return self._toys_loss[parameter.name]
+
+    def _generate_and_fit_toys(self, ntoys, poigen, poieval, printfreq=0.2):
+        """
+        Generate and fit toys for a given POI.
+
+        Args:
+            ntoys (int): number of toys to generate
+            poigen (POI): POI used to generate the toys
+            poieval (POIarray): POI values to evaluate the loss function
+            printfreq: print frequency of the toys generation
+        """
+
+        minimizer = self.minimizer
+
+        param = poigen.parameter
+        gen_value = poigen.value
+
+        toys_loss = self.toys_loss(param)
+        sampler = toys_loss.data
+
+        result = {"bestfit": {"values": np.empty(ntoys), "nll": np.empty(ntoys)},
+                  "nll": {p: np.empty(ntoys) for p in poieval}}
+
+        printfreq = ntoys * printfreq
+
+        toys = self.sample(sampler, int(ntoys*1.2), param, gen_value)
+
+        for i in range(ntoys):
+            converged = False
+            toprint = i % printfreq == 0
+            while converged is False:
+                try:
+                    next(toys)
+                except StopIteration:
+                    to_gen = ntoys - i
+                    toys = self.sample(sampler, int(to_gen*1.2), param, gen_value)
+                    next(toys)
+
+                minimum = minimizer.minimize(loss=toys_loss)
+                converged = minimum.converged
+
+                if not converged:
+                    self.set_dependents_to_bestfit()
+                    continue
+
+                bestfit = minimum.params[param]["value"]
+                result["bestfit"]["values"][i] = bestfit
+                nll = pll(minimizer, toys_loss, POI(param, bestfit))
+                result["bestfit"]["nll"][i] = nll
+
+                for p in poieval:
+                    nll = pll(minimizer, toys_loss, p)
+                    result["nll"][p][i] = nll
+
+            if toprint:
+                print("{0} toys generated, fitted and scanned!".format(i))
+
+            if i > ntoys:
+                break
+            i += 1
+
+        return result
+
+    def _get_toys(self, poigen, poieval=None, qtilde=False, hypothesis="null"):
+        """
+        Return the generated toys for a given POI.
+
+        Args:
+            poigen (POI): POI used to generate the toys
+            poieval (POIarray): POI values to evaluate the loss function
```

Comment:
Add that it's optional. If I understand correctly the toys will be generated and fitted at `poigen` *and* `poieval` values … Worth adding?

Reply:
generated at `poigen` and fitted at `poigen` and `poieval`. I will add a line to precise this :)